### ISSUE TITLE
modules/yumpkg.py  refresh_db uses expire-cache followed by expensive check-update

### ISSUE THREAD

**Issue by damon-atkins** [saltstack/salt#36164]
### Description of Issue/Question

yum execution module for refresh_db calls

```
    clean_cmd = [_yum(), '--quiet', 'clean', 'expire-cache']
    update_cmd = [_yum(), '--quiet', 'check-update']  # optional runs this, but by default it does
```

To

```
    clean_cmd = [_yum(), '--quiet', 'clean', 'expire-cache']
    update_cmd = [_yum(), '--quiet', 'makecache']
```

It should run `makecache` every time, as it re-creates the meta data, and also validates that the repo are configured correctly and reachable.    Not sure about why check-update is needed as it would be more expensive than `makecache`

Given no one has complain about  `expire-cache` it should not need to be change  `clean metadata`.

dnf clean expire-cache
       Removes local cookie files saying when the metadata and mirrorlists were downloaded for each repo. **DNF will re-validate**  the  cache  for each repo the next time it is used.

dnf clean metadata
       Removes repository metadata. Those are the files which DNF uses to determine the remote  vailability of packages. Using this option will **make DNF download** all the metadata the next time it is run.

dnf [options] makecache
       Downloads and caches in binary format metadata for all known repos. Tries to avoid downloading whenever  possible  (e.g.  when  the local metadata hasn't expired yet or when the metadata timestamp hasn't changed).

dnf [options] check-update [<package-specs>...]
   Non-interactively checks if updates of the specified packages are available. If no <package-specs> are given checks whether any updates at  all  are  available  for  your system. DNF exit code will be 100 when **there are updates available and a list of the updates** will be  printed, 0 if not and 1 if an error occurs.    **i.e. note the extra processing involved other than re-creating the metadata cache**
### Versions Report

develop and most likely older versions.

**Comment by Ch3LL** [comment]
@damon-atkins based on the definitions of these I believe you are on the right track but just to confirm ping @terminalmage do you agree with @damon-atkins 's assessment of improvements to the code in this module?

**Comment by damon-atkins** [comment]
```
# time yum --verbose clean expire-cache
Loading "fastestmirror" plugin
Config time: 0.006
Yum version: 3.4.3
Cleaning repos: base extras salt-latest updates
7 metadata files removed

real    0m0.202s
user    0m0.185s
sys     0m0.017s
# time yum  check-update
Loaded plugins: fastestmirror
base                                                                                                                    | 3.6 kB  00:00:00
extras                                                                                                                  | 3.4 kB  00:00:00
salt-latest                                                                                                             | 2.9 kB  00:00:00
updates                                                                                                                 | 3.4 kB  00:00:00
Loading mirror speeds from cached hostfile
 * base: mirror.internode.on.net
 * extras: mirror.internode.on.net
 * updates: mirror.internode.on.net

List of packages which need updating removed

real    0m3.172s
user    0m0.434s
sys     0m0.099s
```

```
# time yum --verbose clean expire-cache
Loading "fastestmirror" plugin
Config time: 0.005
Yum version: 3.4.3
Cleaning repos: base extras salt-latest updates
7 metadata files removed

real    0m0.155s
user    0m0.137s
sys     0m0.018s

# time yum  --verbose makecache
Loading "fastestmirror" plugin
Config time: 0.006
Yum version: 3.4.3
base                                                                                                                    | 3.6 kB  00:00:00
extras                                                                                                                  | 3.4 kB  00:00:00
salt-latest                                                                                                             | 2.9 kB  00:00:00
updates                                                                                                                 | 3.4 kB  00:00:00
Loading mirror speeds from cached hostfile
 * base: mirror.internode.on.net
 * extras: mirror.internode.on.net
 * updates: mirror.internode.on.net
Metadata Cache Created

real    0m2.508s
user    0m0.245s
sys     0m0.087s
```
With a small repo you save about half a second, surprising this smaller difference than I expected.

**Comment by takeda** [comment]
If I may add, I would prefer if salt would not decide for me when to prune cache. When I run salt I see it is doing it few times in one run, it only slows down package installation and defeats the purpose of cache. Most of the time when salt is run there's no change in the repo, so why purge cache?

The proper way is to use metadata_expire setting in yum to configure how long cache is valid, and if cache is needed to be purged sooner, nothing stops from purging cache by hand.

I would like to have this feature removed or at least allow for disabling it.

If someone wants to purge cache on every salt run, they can can use cmd.run to call the command.

**Comment by damon-atkins** [comment]
Their was a bug with state runs, where it was doing a refresh multiple times for all platforms (pkg.install or pkg.latest can not remember which one). It should only do it once per state or highstate.    I reworked the code and the fix is in 2016.11 and  pkg.install or pkg.latest use the new code.

I fixed it to help with the Windows Package management, as refresh was being called be each pkg. install/latest. And on windows refresh is very expensive.

And in looking at how the code worked, I raised this issue.

**Comment by takeda** [comment]
We use 2015.5.3 so I guess it has that bug. In any case even if it didn't I would prefer salt wouldn't do the purge unless I would explicitly tell it to do it (through the module/state or even cmd.run). Current behavior only slows it down (vast majority of times there's no change in repo) and puts an extra load on server hosting the repos.

Please remove this behavior, the cache exists to speed things up, pruning cache on every run defeats its intended purpose.
